tests/validation/src/jacobian_validation.cpp

```diff
@@ -0,0 +1,1475 @@
+    using actor_chain_t = actor_chain<dtuple, propagation::print_inspector,
+                                      parameter_transporter<transform3_type>,
+                                      bound_getter<transform3_type>,
+                                      parameter_resetter<transform3_type>>;
+
+    // Iterate over reference (pilot) tracks for a rectangular telescope
+    // geometry and Jacobian calculation
+    using uniform_gen_t =
+        random_numbers<scalar, std::uniform_real_distribution<scalar>,
+                       std::seed_seq>;
+    using trk_generator_t = random_track_generator<track_type, uniform_gen_t>;
+    trk_generator_t::configuration trk_gen_cfg{};
+    trk_gen_cfg.n_tracks(n_tracks + n_skips);
+    trk_gen_cfg.phi_range(-constant<scalar>::pi + 0.1f,
+                          constant<scalar>::pi - 0.1f);
+    trk_gen_cfg.theta_range(0.1f, constant<scalar>::pi - 0.1f);
+    trk_gen_cfg.mom_range(0.5f * unit<scalar>::GeV, 100.f * unit<scalar>::GeV);
+    trk_gen_cfg.origin({0.f, 0.f, 0.f});
+    trk_gen_cfg.origin_stddev({0.f * unit<scalar>::mm, 0.f * unit<scalar>::mm,
+                               0.f * unit<scalar>::mm});
+
+    unsigned int track_count = 0u;
+
+    // Vectors for dqopdqop relative difference
+    std::vector<std::vector<scalar>> dqopdqop_rel_diffs_rect(log10_tols.size());
+    std::vector<std::vector<scalar>> dqopdqop_rel_diffs_wire(log10_tols.size());
+
+    for (const auto track : trk_generator_t{trk_gen_cfg}) {
+
+        // Navigator types
+        using rect_navigator_t = navigator<rectangle_telescope>;
+        using wire_navigator_t = navigator<wire_telescope>;
+
+        // Stepper types
+        using const_field_stepper_t =
+            rk_stepper<const_bfield_t::view_t, transform3_type,
+                       constrained_step<>, stepper_default_policy>;
+        using inhom_field_stepper_t =
+            rk_stepper<inhom_bfield_t::view_t, transform3_type,
+                       constrained_step<>, stepper_default_policy>;
+
+        // Make four propagators for each case
+        using const_field_rect_propagator_t =
+            propagator<const_field_stepper_t, rect_navigator_t, actor_chain_t>;
+        using inhom_field_rect_propagator_t =
+            propagator<inhom_field_stepper_t, rect_navigator_t, actor_chain_t>;
+
+        using const_field_wire_propagator_t =
+            propagator<const_field_stepper_t, wire_navigator_t, actor_chain_t>;
+        using inhom_field_wire_propagator_t =
+            propagator<inhom_field_stepper_t, wire_navigator_t, actor_chain_t>;
+
+        // Pilot track
+        detail::helix<transform3_type> helix_bz(track, &B_z);
+
+        // Make a telescope geometry with rectagular surface
+        const scalar detector_length = rand_length(mt);
+        tel_det_config<rectangle2D<>, detail::helix<transform3_type>>
+            rectangle_cfg{rect, helix_bz};
+        rectangle_cfg.m_envelope = 1000.f * unit<scalar>::mm;
```

Comment:
I guess this is needed because the surfaces are tilted after portal creation?

Reply:
Surfaces are tilted and shifted indeed. Overstepping can happen as well.

Reply:
There is an exit condition in `bound_getter` when the track reaches the final surface

```
template <typename transform3_t>
struct bound_getter : actor {

    // Transformation matching this struct
    using transform3_type = transform3_t;
    // scalar_type
    using scalar_type = typename transform3_type::scalar_type;
    using bound_track_parameters_type = bound_track_parameters<transform3_t>;
    using free_track_parameters_type = free_track_parameters<transform3_t>;

    struct state {

        scalar m_path_length;
        bound_track_parameters_type m_param_departure;
        bound_track_parameters_type m_param_destination;
        typename bound_track_parameters_type::covariance_type m_jacobi;
        scalar m_avg_step_size{0.f};
        std::size_t step_count{0u};
    };

    template <typename propagator_state_t>
    DETRAY_HOST_DEVICE void operator()(state& actor_state,
                                       propagator_state_t& propagation) const {

        auto& navigation = propagation._navigation;
        auto& stepping = propagation._stepping;

        actor_state.step_count++;

        const scalar N = static_cast<scalar>(actor_state.step_count);

        actor_state.m_avg_step_size =
            ((N - 1.f) * actor_state.m_avg_step_size + stepping._step_size) / N;

        if (navigation.is_on_module() && navigation.barcode().index() == 0u) {

            actor_state.m_param_departure = stepping._bound_params;
        }
        // Get the bound track parameters and jacobian at the destination
        // surface
        else if (navigation.is_on_module() &&
                 navigation.barcode().index() == 1u) {

            actor_state.m_path_length = stepping._path_length;
            actor_state.m_param_destination = stepping._bound_params;
            actor_state.m_jacobi = stepping._full_jacobian;

            // Stop navigation if the destination surface found
            propagation._heartbeat &= navigation.exit();
        }

        return;
    }
};

```